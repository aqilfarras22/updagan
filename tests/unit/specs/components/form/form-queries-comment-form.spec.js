import BootstrapVue from 'bootstrap-vue';
import VeeValidate from 'vee-validate';
import { mount, createLocalVue } from '@vue/test-utils';

import formQueriesCommentFormComponent from '@/components/form/form-queries/form-queries-comment-form';

const { it, describe, expect } = global;

const testPropsData = {
  saveBtnText: 'saveBtnText',
};

const localVue = createLocalVue();
localVue.use(BootstrapVue);
localVue.use(VeeValidate, {
  fieldsBagName: 'veeFields',
});

describe('components/form/form-queries/form-queries-comment-form.vue', () => {
  const createMountedWrapper = (options = {}) => mount(formQueriesCommentFormComponent, {
    localVue,
    propsData: { ...testPropsData },
    stubs: {
      ButtonCmp: {
        name: 'ButtonCmp',
        render: h => h('b', 'ButtonCmp'),
      },
    },
    sync: false,
    ...options,
  });
  it('should create instance', () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  it('should have described child components', () => {
    expect(formQueriesCommentFormComponent.components.ButtonCmp).toBeTruthy();
  });

  it('should able to receive props', () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.vm.fakeProp).toBeUndefined();
    expect(wrapper.vm.saveBtnText).toEqual(testPropsData.saveBtnText);
    expect(wrapper.vm.saveBtnText).not.toBeUndefined();
  });

  it('should have all required initial data', () => {
    const wrapper = createMountedWrapper();
    // initial data
    expect(wrapper.vm.comment).toBe('');
    expect(wrapper.vm.limit).toBe(500);
    expect(wrapper.vm.submited).toBe(false);
  });

  it('should watch comment prop and slice it with limit', async () => {
    const wrapper = createMountedWrapper();
    wrapper.setData({
      comment: '12345678',
      limit: 5,
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.comment).toBe('12345');
    wrapper.setData({
      comment: '12345678',
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.comment).toBe('12345');
    wrapper.setData({
      comment: '1234',
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.comment).toBe('1234');
  });

  it('should have method which emit "cancel" event on comment cancel: cancel', async () => {
    const wrapper = createMountedWrapper();
    const cancelLink = wrapper.find('.cancelBtn');
    expect(cancelLink.is('b')).toBe(true);

    wrapper.vm.cancel();

    // assert event has been emitted
    expect(wrapper.emitted().cancel).toBeTruthy();
    // assert event count
    expect(wrapper.emitted().cancel.length).toBe(1);
  });

  it('should have method which emit "submit" event on comment submit (if valid): submit', async () => {
    const wrapper = createMountedWrapper();
    expect(wrapper.vm.submited).toBe(false);

    try {
      await wrapper.vm.submit();
    } catch (e) {
      expect(e.message).toMatch('Invalid comment data');
    }
    expect(wrapper.vm.submited).toBe(true);

    // invalid from data
    expect(wrapper.emitted().submit).toBeFalsy();

    const commentField = wrapper.find('#comment');
    expect(commentField.is('textarea')).toBe(true);
    commentField.element.value = 'My comment';
    commentField.trigger('input');
    await wrapper.vm.$nextTick();
    await wrapper.vm.submit();

    // assert event has been emitted
    expect(wrapper.emitted().submit).toBeTruthy();
    // assert event count
    expect(wrapper.emitted().submit.length).toBe(1);
    // assert event payload
    expect(wrapper.emitted().submit[0][0]).toEqual({ comment: 'My comment' });
  });

  it('has the expected html structure', async () => {
    const wrapper = createMountedWrapper();
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$el).toMatchSnapshot();

    try {
      await wrapper.vm.submit();
    } catch (e) {
      expect(e.message).toMatch('Invalid comment data');
    }
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.$el).toMatchSnapshot();

    const commentField = wrapper.find('#comment');
    expect(commentField.is('textarea')).toBe(true);
    commentField.element.value = 'My comment';
    commentField.trigger('input');
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.$el).toMatchSnapshot();
  });
});
