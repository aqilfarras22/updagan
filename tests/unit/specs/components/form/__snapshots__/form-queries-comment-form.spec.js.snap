// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/form/form-queries/form-queries-comment-form.vue has the expected html structure 1`] = `
<form
  class="component"
>
  <div
    class="form-row"
  >
    <fieldset
      class="group b-form-group form-group"
      id="__BVID__39"
      role="group"
    >
      <!---->
      <div
        class=""
        role="group"
      >
        <textarea
          aria-invalid="false"
          aria-required="true"
          class="form-control"
          id="comment"
          name="comment"
          placeholder="Enter your comment"
          rows="6"
          wrap="soft"
        />
         
        <span
          class="note"
        >
          Max allowed limit 500 chars
        </span>
        <!---->
        <!---->
        <!---->
      </div>
    </fieldset>
     
    <div
      class="form-row"
    >
      <div
        class="col"
      >
        <b
          type="submit"
          variant="primary"
        >
          ButtonCmp
        </b>
         
        <b
          class="ml-2 cancelBtn"
          variant="outline-primary"
        >
          ButtonCmp
        </b>
      </div>
    </div>
  </div>
</form>
`;

exports[`components/form/form-queries/form-queries-comment-form.vue has the expected html structure 2`] = `
<form
  class="component"
>
  <div
    class="form-row"
  >
    <fieldset
      aria-describedby="__BVID__39__BV_feedback_invalid_"
      aria-invalid="true"
      class="group b-form-group form-group is-invalid"
      id="__BVID__39"
      role="group"
    >
      <!---->
      <div
        class=""
        role="group"
      >
        <textarea
          aria-invalid="true"
          class="form-control is-invalid"
          id="comment"
          name="comment"
          placeholder="Enter your comment"
          rows="6"
          wrap="soft"
        />
         
        <span
          class="note"
        >
          Max allowed limit 500 chars
        </span>
        <div
          aria-atomic="true"
          aria-live="assertive"
          class="invalid-feedback d-block"
          id="__BVID__39__BV_feedback_invalid_"
          role="alert"
        >
          The comment field is required.
        </div>
        <!---->
        <!---->
      </div>
    </fieldset>
     
    <div
      class="form-row"
    >
      <div
        class="col"
      >
        <b
          type="submit"
          variant="primary"
        >
          ButtonCmp
        </b>
         
        <b
          class="ml-2 cancelBtn"
          variant="outline-primary"
        >
          ButtonCmp
        </b>
      </div>
    </div>
  </div>
</form>
`;

exports[`components/form/form-queries/form-queries-comment-form.vue has the expected html structure 3`] = `
<form
  class="component"
>
  <div
    class="form-row"
  >
    <fieldset
      aria-describedby="__BVID__39__BV_feedback_invalid_"
      aria-invalid="true"
      class="group b-form-group form-group is-invalid"
      id="__BVID__39"
      role="group"
    >
      <!---->
      <div
        class=""
        role="group"
      >
        <textarea
          aria-invalid="true"
          class="form-control is-invalid"
          id="comment"
          name="comment"
          placeholder="Enter your comment"
          rows="6"
          wrap="soft"
        />
         
        <span
          class="note"
        >
          Max allowed limit 500 chars
        </span>
        <div
          aria-atomic="true"
          aria-live="assertive"
          class="invalid-feedback d-block"
          id="__BVID__39__BV_feedback_invalid_"
          role="alert"
        >
          The comment field is required.
        </div>
        <!---->
        <!---->
      </div>
    </fieldset>
     
    <div
      class="form-row"
    >
      <div
        class="col"
      >
        <b
          type="submit"
          variant="primary"
        >
          ButtonCmp
        </b>
         
        <b
          class="ml-2 cancelBtn"
          variant="outline-primary"
        >
          ButtonCmp
        </b>
      </div>
    </div>
  </div>
</form>
`;
